service/agentconf.go

```diff
@@ -0,0 +1,274 @@
+}
+
+// CreateAgentConf creates the configfile for specified agent running on a host with specified series.
+func CreateAgentConf(agentName string, dataDir string, series string) (_ common.Conf, err error) {
+	defer func() {
+		if err != nil {
+			errors.Errorf("failed create agent conf for %s: %s", agentName, err)
+		}
+	}()
+
+	renderer, err := shell.NewRenderer("")
+	if err != nil {
+		return common.Conf{}, err
+	}
+
+	tag, err := names.ParseTag(agentName)
+	if err != nil {
+		return common.Conf{}, err
+	}
+	name := tag.Id()
+
+	var kind AgentKind
+	switch tag.Kind() {
+	case names.MachineTagKind:
+		kind = AgentKindMachine
+	case names.UnitTagKind:
+		kind = AgentKindUnit
+	default:
+		return common.Conf{}, errors.NewNotValid(nil, fmt.Sprintf("agent %q is neither a machine nor a unit", agentName))
+	}
+
+	info := NewAgentInfo(
+		kind,
+		name,
+		dataDir,
+		paths.MustSucceed(paths.LogDir(series)),
+	)
+	return AgentConf(info, renderer), nil
+}
+
+// CopyAgentBinary copies all the tools into the path specified for each agent.
+func CopyAgentBinary(machineAgent string, unitAgents []string, dataDir string, toSeries string, fromSeries string, jujuVersion version.Number) (err error) {
+	defer func() {
+		if err != nil {
+			errors.Annotate(err, "failed to copy tools")
+		}
+	}()
+
+	// Setup new and old version.Binarys with only the series
+	// different.
+	fromVers := version.Binary{
+		Number: jujuVersion,
+		Arch:   arch.HostArch(),
+		Series: fromSeries,
+	}
+	toVers := version.Binary{
+		Number: jujuVersion,
+		Arch:   arch.HostArch(),
+		Series: toSeries,
+	}
+
+	// If tools with the new series don't already exist, copy
+	// current tools to new directory with correct series.
+	if _, err = os.Stat(tools.SharedToolsDir(dataDir, toVers)); err != nil {
+		// Copy tools to new directory with correct series.
+		if err = fs.Copy(tools.SharedToolsDir(dataDir, fromVers), tools.SharedToolsDir(dataDir, toVers)); err != nil {
+			return err
+		}
+	}
+
+	// Write tools metadata with new version, however don't change
+	// the URL, so we know where it came from.
+	jujuTools, err := tools.ReadTools(dataDir, toVers)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Only write once
+	if jujuTools.Version != toVers {
+		jujuTools.Version = toVers
+		if err = tools.WriteToolsMetadataData(tools.ToolsDir(dataDir, toVers.String()), jujuTools); err != nil {
+			return err
+		}
+	}
+
+	// Update Agent Tool links
+	var lastError error
+	for _, agentName := range append(unitAgents, machineAgent) {
+		toolPath := tools.ToolsDir(dataDir, toVers.String())
+		toolsDir := tools.ToolsDir(dataDir, agentName)
+
+		err = symlink.Replace(toolsDir, toolPath)
+		if err != nil {
+			lastError = err
+		}
+	}
+
+	return lastError
+}
+
+// StartAllAgents starts all the agents in the machine with specified series.
+func StartAllAgents(machineAgent string, unitAgents []string, dataDir string, series string) (errMachineName string, errUnitNames []string, err error) {
```

Comment:
Same here - these should be local vars rather than named returns,  `errMachineName` should be `machine` or `machineName`, `errUnitNames` should be `units` or `unitNames`.